sequencer/src/api.rs

```diff
@@ -3389,6 +3393,213 @@ mod test {
+        for block in 0..=EPOCH_HEIGHT * 2 {
+            for address in addresses.clone() {
+                let amount = client
+                    .get::<Option<RewardAmount>>(&format!(
+                        "reward-state/reward-balance/{block}/{address}"
+                    ))
+                    .send()
+                    .await
+                    .ok()
+                    .flatten();
+                assert!(amount.is_none(), "amount is not none for block {block}")
+            }
+        }
+
+        // Collect leaves for epoch 3 to 5 to verify reward calculations.
+        let leaves = client
+            .socket("availability/stream/leaves/41")
+            .subscribe::<LeafQueryData<SeqTypes>>()
+            .await
+            .unwrap()
+            .take((EPOCH_HEIGHT * 3).try_into().unwrap())
+            .try_collect::<Vec<_>>()
+            .await
+            .unwrap();
+
+        let node_state = network.server.node_state();
+        let coordinator = node_state.coordinator;
+
+        let mut rewards_map = HashMap::new();
+
+        for leaf in leaves {
+            let block = leaf.height();
+            tracing::info!("verify rewards for block={block:?}");
+            let membership = coordinator.membership().read().await;
+            let epoch = epoch_from_block_number(block, EPOCH_HEIGHT);
+            let epoch_number = EpochNumber::new(epoch);
+            let leader = membership
+                .leader(leaf.leaf().view_number(), Some(epoch_number))
+                .expect("leader");
+            let leader_eth_address = membership.address(&epoch_number, leader).expect("address");
+            drop(membership);
+
+            let validators = client
+                .get::<IndexMap<Address, Validator<BLSPubKey>>>(&format!("node/validators/{epoch}"))
+                .send()
+                .await
+                .expect("validators");
+
+            let leader_validator = validators
+                .get(&leader_eth_address)
+                .expect("leader not found");
+
+            // Verify that the sum of delegator stakes equals the validator's total stake.
+            for validator in validators.values() {
+                let delegator_stake_sum: U256 = validator.delegators.values().cloned().sum();
+
+                assert_eq!(delegator_stake_sum, validator.stake);
+            }
+
+            let rewards = compute_rewards(leader_validator.clone()).expect("reward computation");
+
+            // Verify that the leader commission amount is within the tolerated range.
+            // Due to potential rounding errors in decimal calculations for delegator rewards,
+            // the actual distributed commission
+            // amount may differ very slightly from the calculated value.
+            // this asserts that it is within 10wei tolerance level.
+            // 10 wei is 10* 10E-18
+            let (_, leader_commission_amount) = rewards.last().unwrap();
```

Comment:
I think we shouldn't depend on the order in which rewards are returned. I don't think it should be a requirements of the the compute_rewards function to put the leader reward last. IIUC currently if the leader is also a delegator we will have 2 entries for rewards that go to the leader.

Reply:
How about adding a type for rewards instead of using a `Vec`? With public methods like `leader_reward` that are safe to use. I think the vec is easy to misuse. Like it's tempting to assume that it only contains each address once but it won't if the leader is also a delegator.

Reply:
This vec actually contains two entries for leaders. one is for commission reward and other the delegation reward. I just wanted the commission one so that seemed the easiest in this test. I don't know how to differentiate between these two entries. Maybe I should let compute_rewards return commission separately as well

Reply:
I mean if we have our own type we can add a function to it to get the leader reward.

Reply:
I just think we should encapsulate this logic rather than having an implicit dependency on what order things are put in a vec and I think this is most easily solved by creating a new type that we return from compute rewards instead of using a vec.

Reply:
ok I'll add it